feat: substitute store paths in parallel

zmitchell opened the pull request:
## Proposed Changes

<!-- Describe the changes proposed in this pull request. -->
<!-- Please provide links to any issue(s) which are expected to be resolved. -->

This change refactors `buildenv.rs` to substitute each package in parallel with a configurable number of maximum parallel downloads.

Notable changes:
- Realizing a lockfile now does so via package type specific functions e.g. `realise_single_base_catalog_pkg`.
- These functions are static methods on `BuildEnvNix` now instead of instance methods because the generic types on `BuildEnvNix<P, A>` in the impl block aren't constrained in such a way that requires them to be thread safe. This was also an opportunity to disentangle some of the download logic for base catalog packages and custom catalog packages.
- We now do as much prep ahead of time as possible so that we can essentially loop over all packages and spawn threads for each one.
- We use a semaphore to control the maximum number of _active_ downloads, but we still spawn a thread for each package at the beginning (and a user will still see spinners for all packages). Preventing threads from being spawned until they're ready to download isn't worth the complexity in my opinion.

## Release Notes

<!-- Describe any user facing changes. Use "N/A" if not applicable. -->
Packages will now be downloaded in parallel when building environments. The maximum number of parallel downloads can be controlled via the new `FLOX_MAX_PARALLEL_DOWNLOADS` variable.

<!-- Many thanks! -->

ysndr reviewed (approved):
> You don't get a spinner per package in that case without some gymnastics

I think gymnastics are the way to go eventually just to have consistent logs from all nix based processes (gc, substitution, env builds, package builds, containerize, ...).
I concede that nix logs will likely be more verbose, and expose the runtime closure to users (via downloading runtime dependencies) -- I'm unsure if that's a good or a bad thing.

> I don't think setting that on individual processes will contribute towards a global limit, which is what a user needs.
> This could work if we do all substitutions in a single nix build, but I don't think is a good idea due to how we handle catalog packages.

I was thinking of sequentially batching packages by substituter, but i see how thats #fun with custom packages potentially having multiple locations.

Besides, for every parallel nix substitution, nix itself might spawn a number of threads to fetch dependencies in parallel.

Iirc parallelism is also fixed at the daemon level. In that case you could spawn abitrarily many `nix copy` processes but they'd still be blocked by the daemon.
I've tried converting our existing loops into `rayon::par_iter` loops in the past and did _not_ see huge performance liftings.

> It also wouldn't really let us have meaningful spinners, they're either all going or all done.

I think nix also emits spans of some sort (though those would be store path based rather than package based).


> Re: more things to break:
>    Since I just reorganized the functionality that already exists, as far as I can tell the biggest risk is that a thread panics during join. If that happens it's highly unlikely that it's our fault or that there isn't a bigger problem to deal with.

Yes, this looked like more change than it is indeed.

-------

bottom line is, given that parsing nix logs is a bit awkward and that we dont know which location we want to fetch custom packages from, i agree this is probably the closest we get by ourselves.
I'm not entirely conviced that it will have a huge impact if the nix daemon sequentializes things in the end though.

zmitchell commented on cli/flox-rust-sdk/src/providers/buildenv.rs:
```diff
@@ -515,54 +652,34 @@ where
+                return Err(BuildEnvError::Other(format!(
+                    "Locked package '{}' is a base catalog package, but the locked url '{}' does not start with the expected prefix '{}'",
+                    locked_pkg.install_id, locked_pkg.locked_url, NIXPKGS_CATALOG_URL_PREFIX
+                )));
```
I'm not intending to assign significance here. A different name for this could be "custom."

zmitchell commented on cli/flox-rust-sdk/src/providers/buildenv.rs:
```diff
@@ -292,221 +332,318 @@ where
-    /// IMPORTANT/TODO: As custom catalogs, with non-nixpkgs packages are in development,
-    /// this function is currently assumes that the package is from the nixpkgs base-catalog.
-    /// Currently the type is distinguished by the [LockedPackageCatalog::locked_url].
-    /// If this does not indicate a nixpkgs package, the function will currently panic!
```
Yeah I have it queued up locally

zmitchell commented on cli/flox-rust-sdk/src/providers/buildenv.rs:
```diff
@@ -292,221 +332,318 @@ where
+        let semaphore = if let Some(max_par) = max_parallel_downloads {
+            Semaphore::new(max_par, max_par)
+        } else {
+            Semaphore::new(u16::MAX, u16::MAX)
```
I've reduced this to 20 just to get this over the line. I'm still not super concerned about it in practice though.

zmitchell commented:
I've addressed all of the comments except https://github.com/flox/flox/pull/3785#discussion_r2641391998, which I think is a matter of personal preference, so I'm going ahead and merging.
